src/coreclr/src/vm/genericdict.cpp

```diff
@@ -109,194 +108,343 @@ DictionaryLayout::GetFirstDictionaryBucketSize(
+            _ASSERT(SystemDomain::SystemModule()->m_DictionaryCrst.OwnedByCurrentThread());
+
+            PVOID pResultSignature = pSigBuilder == NULL ? pSig : CreateSignatureWithSlotData(pSigBuilder, pAllocator, slot);
+            pDictLayout->m_slots[iSlot].m_signature = pResultSignature;
+            pDictLayout->m_slots[iSlot].m_signatureSource = signatureSource;
+
+            pResult->signature = pDictLayout->m_slots[iSlot].m_signature;
+
+            _ASSERTE(FitsIn<WORD>(nFirstOffset + 1));
+            pResult->indirections = static_cast<WORD>(nFirstOffset + 1);
+            pResult->offsets[nFirstOffset] = slot * sizeof(DictionaryEntry);
+            *pSlotOut = slot;
+            return TRUE;
         }
 
-        // If we've reached the end of the chain we need to allocate another bucket. Make the pointer update carefully to avoid
-        // orphaning a bucket in a race. We leak the loser in such a race (since the allocation comes from the loader heap) but both
-        // the race and the overflow should be very rare.
-        if (pDictLayout->m_pNext == NULL)
-            FastInterlockCompareExchangePointer(&pDictLayout->m_pNext, Allocate(4, pAllocator, NULL), 0);
+        slot++;
+    }
 
-        pDictLayout = pDictLayout->m_pNext;
-        isFirstBucket = FALSE;
+    *pSlotOut = pDictLayout->m_numSlots;
+    return FALSE;
+}
+
+#ifndef CROSSGEN_COMPILE
+/* static */
+DictionaryLayout* DictionaryLayout::ExpandDictionaryLayout(LoaderAllocator*                 pAllocator, 
+                                                           DictionaryLayout*                pCurrentDictLayout, 
+                                                           DWORD                            numGenericArgs, 
+                                                           SigBuilder*                      pSigBuilder, 
+                                                           BYTE*                            pSig, 
+                                                           int                              nFirstOffset, 
+                                                           DictionaryEntrySignatureSource   signatureSource, 
+                                                           CORINFO_RUNTIME_LOOKUP*          pResult,
+                                                           WORD*                            pSlotOut)
+{
+    CONTRACTL
+    {
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(ThrowOutOfMemory(););
+        PRECONDITION(SystemDomain::SystemModule()->m_DictionaryCrst.OwnedByCurrentThread());
+        PRECONDITION(CheckPointer(pResult) && CheckPointer(pSlotOut));
     }
-} // DictionaryLayout::FindToken
+    CONTRACTL_END
+        
+    // There shouldn't be any empty slots remaining in the current dictionary.
+    _ASSERTE(pCurrentDictLayout->m_slots[pCurrentDictLayout->m_numSlots - 1].m_signature != NULL);
+
+#ifdef _DEBUG
+    // Stress debug mode by increasing size by only 1 slot
```

Comment:
This will turn linear algorithm into quadratic algorithm for debug builds. Should we limit the amount of stress we do in debug builds? E.g. by growing by 1 only up to 10 slots, and the switch to doubling?

Reply:
We could. I've had the +1 suggested in the original PR for debug builds as a way to stress the feature :)